src/firefly/js/drawingLayers/Catalog.js

```diff
@@ -0,0 +1,182 @@
+var createCnt= 0;
+const defColors= [COLOR_PT_1, COLOR_PT_2, COLOR_PT_3, COLOR_PT_5, COLOR_PT_6];
+
+
+function creator(initPayload) {
+    const {catalogId, tableData, tableMeta, columns, tableRequest, highlightedRow}= initPayload;
+    var drawingDef= makeDrawingDef();
+    drawingDef.symbol= DrawSymbol.SQUARE;
+
+    var pairs= {
+        [MouseState.DOWN.key]: highlightChange
+    };
+
+    drawingDef.color= tableMeta[MetaConst.DEFAULT_COLOR] || defColors[createCnt % defColors.length];
+
+    var options= {
+        hasPerPlotData:false,
+        isPointData:true,
+        canUserDelete: true,
+        canUseMouse:true,
+        canHighlight: true,
+        canSelect: true,
+        canFilter: true,
+        helpLine : helpText,
+        canUserChangeColor: ColorChangeType.STATIC
+    };
+    // todo: get the real title
+    const dl= DrawLayer.makeDrawLayer(catalogId,TYPE_ID, `Catalog: ${tableMeta.title || catalogId}`,
+        options, drawingDef, null, pairs );
+    dl.catalogId= catalogId;
+    dl.tableData= tableData;
+    dl.tableMeta= tableMeta;
+    dl.tableRequest= tableRequest;
+    dl.highlightedRow= highlightedRow;
+    dl.columns= columns;
+
+
+    createCnt++;
+    return dl;
+}
+
+
+
+function highlightChange(mouseStatePayload) {
+    const {drawLayer,plotId,screenPt}= mouseStatePayload;
+    makeHighlightDeferred(drawLayer,plotId,screenPt);
+}
+
+
+function makeHighlightDeferred(drawLayer,plotId,screenPt) {
+    var done= false;
+    var idx= 0;
+    const maxChunk= 1000;
+    var minDist = Number.MAX_SAFE_INTEGER;
+    const {data}= drawLayer.drawData;
+    const {tableRequest}= drawLayer;
+    var closestIdx;
+    const plot= primePlot(visRoot(),plotId);
+    const id= window.setInterval( () => {
+        if (done) {
+            window.clearInterval(id);
+            dispatchTableHighlight(drawLayer.drawLayerId,closestIdx,tableRequest);
+        }
+
+        var dist;
+
+        for(var i=0;(idx<data.length && i<maxChunk ); i++) {
+            const obj= data[idx];
+            if (obj) {
+                dist = DrawOp.getScreenDist(obj, plot, screenPt);
+                if (dist > -1 && dist < minDist) {
+                    minDist = dist;
+                    closestIdx= idx;
+                }
+            }
+            idx++;
+        }
+        done= (idx===data.length);
+    },0);
+    return () => window.clearInterval(id);
+}
+
+function getDrawData(dataType, plotId, drawLayer, action, lastDataRet) {
+
+    const{tableData, columns}= drawLayer;
+    switch (dataType) {
+        case DataTypes.DATA:
+            return isEmpty(lastDataRet) ? computeDrawLayer(tableData, columns) : lastDataRet;
+            break;
+        case DataTypes.HIGHLIGHT_DATA:
+            return isEmpty(lastDataRet) ? 
+                          computeHighlightLayer(tableData, drawLayer.highlightedRow, columns) : lastDataRet;
+            break;
+        case DataTypes.SELECTED_IDX_ARY:
+            return null;
+            break;
+    }
```

Comment:
Do we want to return null or throw an exception so it can be re-thrown and trace upfront? Or do we handle the 'null' on the component view after dispatch?